### PR TITLE
chore: implement GetSlot, GetSlots, and GetAgent for K8s rm

### PR DESCRIPTION
**Pull request by amandavialva01**
## Description

Implement `GetSlot`, `GetSlots`, and `GetAgent` for Kubernetes resource manager API.

## Test Plan

No manual testing needed.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

DET-9919, DET-9921

### CODE REVIEW

**Review by carolinaecalderon — APPROVED** [review]

**Comment by stoksc** [comment]
review incoming soon, fyi

File: master/internal/rm/kubernetesrm/pods.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/pods.go:1249]: style nit: https://github.com/uber-go/guide/blob/master/style.md#use-var-for-zero-value-structs, in most of the declarations in this change. to me, its nice to visually distinguish zero val structs, slices and such very clearly from non zero val. take it or leave it, not really followed elsewhere, but i follow this.

so this would be
```suggestion
	var response apiv1.GetSlotsResponse
	for _, slot := range slots {
		response.Slots = append(response.Slots, slot)
	}
	return &response
```

Comment (stoksc) [reply master/internal/rm/kubernetesrm/pods.go:1249]: but really i'd just change this entire block to
```suggestion
	return &apiv1.GetSlotsResponse{Slots: slots}
```

File: master/internal/rm/kubernetesrm/pods.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/pods.go:1258]:
```suggestion
	return  &apiv1.GetSlotResponse{Slot: slot}
```

Comment (stoksc) [reply master/internal/rm/kubernetesrm/pods.go:1258]: this applies in a lot of places, but i'll stop commenting it everywhere and leave you to fix the others.

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:125]: similar to other comments,
```suggestion
	podHandlers := map[string]*pod{
		string(pod1.allocationID)]: pod1,
		string(pod2.allocationID)]: pod2
	}
```

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:137]: just init these in the struct literal above (? i think that is technically the right word to call it: https://go.dev/ref/spec#Composite_literals)

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:143]: same comment about zero val decl

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:56]: use table driven tests for this; it is exactly what it is for. (see https://hpe-aiatscale.atlassian.net/wiki/spaces/ENGINEERIN/pages/1389035562/Writing+Tests+for+Backend+Code#Resources)

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:102]: just make this a separate test

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:113]:
```suggestion
	return k8sV1.Node{
		ObjectMeta: metaV1.ObjectMeta{
			ResourceVersion: "1",
			Name:            name,
		},
		Status: k8sV1.NodeStatus{Allocatable: createMockResourceList()},
	}
```

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:175]: use a map literal here and elsewhere, too

**Review by stoksc — COMMENTED** [review]
this looks great, just assorted style commentary for you

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (stoksc) [reply master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:113]: thought i commented this but i dont see it, so maybe i missed. for a lot of these helpers, i dont think the LoC saved is worth the indirection when someone is reading it. I would just inline the mock agent, resource list stuff.

File: master/internal/rm/kubernetesrm/pods.go
Comment (amandavialva01) [reply master/internal/rm/kubernetesrm/pods.go:1249]: `slots` is a map, but `response.Slots` is a slice!

Comment (stoksc) [reply master/internal/rm/kubernetesrm/pods.go:1249]: ah yeah, then 
```suggestion
	return &apiv1.GetSlotsResponse{Slots: maps.Values(slots)}
```

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (carolinaecalderon) [reply master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go]: @stoksc this reminds me of https://hpe-aiatscale.atlassian.net/browse/DET-9967 -- if you're close to completing this PR, could be useful to merge that first & then use it here.

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (carolinaecalderon) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:83]: nit (style): match the style of a table-driven test following this example https://golang.cafe/blog/golang-table-test-example#:~:text=Golang%20Table%20Driven%20Testing%20Example
or search for `cases := []struct {` in our code base & you'll see lots of examples. 
Your for loop looks great, it's just a matter of style preference

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (carolinaecalderon) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:124]: same nit here: could replace `tests := map[string] struct{...}` with `cases := []struct {}`

**Review by carolinaecalderon — APPROVED** [review]
LGTM, just a minor style nit & check if [Bradley's ticket](https://hpe-aiatscale.atlassian.net/browse/DET-9967) could be relevant here (if not, disregard).

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (amandavialva01) [reply master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go]: I think helper functions make the test unnecessarily hard to read. Errors should be clear enough with the first layer of tests using `t.Run(...)`, so I removed the nested testing!

Comment (carolinaecalderon) [reply master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go]: Looks better! Still don't like 3 nested loops tbh, @stoksc any thoughts?

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:37]: this doesn't need to be a global var.

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:59]: why do the aux and compute nodes both look like they have GPUs? i know this is a test, but it is strange.

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (carolinaecalderon) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:31]: nit: why not define this as an int64 -- because in the two places you use it, you cast it into that type? (I don't remember if this is allowed in golang literals, if not disregard).

Comment (carolinaecalderon) [reply master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:31]: same thing for `nodeNumSlots`

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (carolinaecalderon) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:28]: nit: only use this variable once, so doesn't need to be a const.

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (carolinaecalderon) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:25]: nit: only use this once, don't need a const -- can substitute instead.

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (carolinaecalderon) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:106]: nit: switch the order of arguments in this. `require.Equal(t, expected, actual)`. I think the expected should be `agent.Id`, right?

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (carolinaecalderon) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:111]: style: lol can we write this as `count := 0` instead ? I don't think it needs to be abbreviated. Even better, since it doesn't seem like we increment it anywhere, and we only call it twice, I think we can just substitute it in.

**Review by carolinaecalderon — COMMENTED** [review]
few style nits

**Comment by stoksc** [comment]
I'm good on this PR, I'll let @carolinaecalderon have the final review and approval.

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (amandavialva01) [reply master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:28]: I use this two places in the test, so for clarity of the test case I'll probably keep this!

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (amandavialva01) [reply master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:111]: Ooh good catch! I incremented this initially but somehow removed that from this commit, I'll change the var name to `count` and add back the incrementation.

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (amandavialva01) [reply master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:220]: We get a nil dereference using things like `podsClientSet.On`, `coveV1Interface.On`, etc. with zero val decl here.

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (amandavialva01) [reply master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go]: Refactored this, added more cluster states, removed loops to test individual functions, and added more test cases!

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go
Comment (carolinaecalderon) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager_intg_test.go:1]: blocking: sorry to do this so late in the game, but I think we should modularize the test better. Instead of having one large `func TestGetAgentsAndGetSlots(t *testing.T) {...}` that includes _all_ tests like `GetSlot, GetAgent, GetAgents etc.`, we should break this up into smaller tests like `func TestGetAgent{...}, func TestGetAgents{...}, func TestGetSlot{...}`. Create a helper function called `setup` that you can run at the start of each test function that properly creates the variables you need.
This [post](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) describes table-driven tests in golang, and note that each test only has 1 "table". I think this is better design because as the test suite grows, developers can follow precedent and add their own 1-table test versus adding more code to `TestGetAgentsAndGetSlots`. Additionally, if the code breaks somewhere, putting each table in their own test will label the failure with that test name, so its even more obvious from CircleCI which function is buggy.

**Review by carolinaecalderon — CHANGES_REQUESTED** [review]

**Review by carolinaecalderon — APPROVED** [review]
LGTM, nice work!

File: master/internal/rm/kubernetesrm/pods.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/pods.go:1256]: logs usually start with lowercase letters

File: master/internal/rm/kubernetesrm/pods.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/pods.go:1285]: we should todo returning an error here. it's just annoying to change the interface since it is also in EE.

**Review by stoksc — APPROVED** [review]

File: master/internal/rm/kubernetesrm/pods.go
Comment (amandavialva01) [reply master/internal/rm/kubernetesrm/pods.go]: made a ticket for this and added the todo!

**Pull request merged**
